### DISCUSSION TITLE
Separate volume controls for sound effects & music?

### DISCUSSION THREAD

**Discussion by Beastman1976** [djyt/cannonball#160] [Ideas]
Hi, I have managed to add my own wav files, but the main problem I'm having is the sound effects of the game (tyre squeel etc) are way too loud compared to the music. Is there any way to lower the volume of the game sound effects? It is the game that is too loud, not the music that is too quiet.

I can boost the volume of the wav file in audacity, but is not ideal as it ruins the quality of the music. To get it to the right volume level means peaking the levels way higher than they should be. Not sure what the technical term for that is, but I know it's a big no no in music production.

In other versions of outrun I have played, separate volume controls for music and effects have been standard in the game menu.

Any help appreciated...

**Comment by J1mbo** [comment]
The wave files are halved in volume and then mixed with equal weighting with the other (PCM and Yamaha) sounds in void Audio::tick() in src/main/sdl2/audio.cpp:



    // And mix them into the mix_buffer
    for (int i = 0; i < samples_written; i++)
    {
        int32_t mix_data = wav_buffer[wavfile.pos] + pcm_buffer[i] + ym_buffer[i];

        // Clip mix data
        if (mix_data > INT16_MAX) mix_data = INT16_MAX;
        else if (mix_data < INT16_MIN) mix_data = INT16_MIN;

        mix_buffer[i] = mix_data;

        // Loop wav files
        if (++wavfile.pos >= wavfile.length)
            wavfile.pos = 0;
    }

To alter the music volume, try appling a multiplier to it e.g.:

`int32_t mix_data = (wav_buffer[wavfile.pos] * 1.5) + pcm_buffer[i] + ym_buffer[i];`

This could be added to the menu and referenced via config.audio for general use-case I guess.

**Comment by Beastman1976** [comment]
Hi, thanks for responding. 

As you seem to be the only guy replying on here with the technical know how, I was wandering if you could help with a different problem? 

When I'm creating games for cannonball using the layout editor, the tyre screech is back to front. So when you go around a corner the screeching only happens if you turn into an easy bend, instead of when you are sliding outwards around a tight bend. Hope that makes sense. 

I'm assuming this should be an easy fix in the code somewhere, maybe by just adding a minus or plus symbol somewhere? It makes a big difference to the gameplay. Any help appreciated. Seems impossible to email Chris about it, but I would really love to get this one annoying bug fixed.

**Reply by J1mbo** [reply]
@Beastman1976 - I've not looked at layout editor I'm afraid. I did though add custom music volume adjustment to my fork (and also mp3 support).

**Comment by Beastman1976** [comment]
I'm only using a windows 10 laptop to run cannonball. I don't think your fork runs on it? Or I can't figure out how to install it anyway. 

This is one of the stages I have made with layout (as are all the other videos on my channel):

https://www.youtube.com/shorts/hq8l0JZ-naA

You can see what I mean about the tyre screech not working properly on bends. The only place in the game it seems to work correctly is when turning left during the split section between stages. The same bends that do not work properly in the original game. Something in the code is in reverse.

**Reply by J1mbo** [reply]
I noticed that no tyre slip is generated when joining a stage on the left lane (so exited previous stage on the right) but it is when joining on the right. The game flips the lanes during the stage transition, so that bug is likely related to that. The LayOut bug is probably not the same.

**Reply by J1mbo** [reply]
We can probably mask this by flipping the decision logic for LayOut stages. I spent some time adding Windows support this week by the way - I've only tested Windows 11 though.

**Comment by Beastman1976** [comment]
That would be cool if it ran on windows, I would definitely give it a go. I'm not a coder, but do like designing stages in layout.

**Reply by J1mbo** [reply]
@Beastman1976 - I had an idea, which is to use the game traction logic combined with the steering wheel angle, rather than try to work out every permutation of flipped roads etc. This seems to work, and should work on all your levels too.

It does change the game in a subtle way. The slip now follows the same logic that the game use to create drift in the first place. So, when fix_bugs is 0 it uses the original logic instead.

I've also had a go at creating a Windows release, i.e. a .exe download (which may or may not work). I'd be interested to know if you can get it to work and if it solves your LayOut stage issues.

**Comment by Beastman1976** [comment]
Hi, thanks for doing this. Really appeciate it. I'm having problems getting it to run though. I've put the roms in the correct directory and installed the VC_redist.x64 file, but still no joy.

I do have Visual Studio installed on my laptop for some reason. When I run debug I'm getting this, if it helps:

'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\cannonball-se.exe'. Module was built without symbols.
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\ntdll.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\kernel32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\KernelBase.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\user32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\win32u.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\gdi32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\gdi32full.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\tinyxml2.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\msvcp_win.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\libGLESv2.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbase.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\shell32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\mpg123.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\shlwapi.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\msvcrt.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\dinput8.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\advapi32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\sechost.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\msvcp140.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\rpcrt4.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\bcrypt.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\vcomp140.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\SDL2.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\imm32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140_1.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\msvcp140_atomic_wait.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\ole32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\combase.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\oleaut32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\setupapi.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\cfgmgr32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE\zlib1.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\winmm.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\dxgi.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\version.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\d3d9.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\kernel.appcore.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\windows.storage.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\wldp.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\dwmapi.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\SHCore.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\InputHost.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\propsys.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\CoreMessaging.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\ws2_32.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\CoreUIComponents.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\WinTypes.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\ntmarta.dll'. 
'cannonball-se.exe' (Win32): Loaded 'C:\Windows\System32\bcryptprimitives.dll'. 
The thread 0x471c has exited with code 1 (0x1).
The thread 0x3274 has exited with code 1 (0x1).
The thread 0x49f4 has exited with code 1 (0x1).
The thread 0x294c has exited with code 1 (0x1).
The thread 0x494c has exited with code 1 (0x1).
The program '[6784] cannonball-se.exe' has exited with code 1 (0x1).

**Reply by J1mbo** [reply]
Sorry this isn't working - it must be frustrating. My guess is that the graphics initialisation is failing. In VS it should have produced some console output, if not start it from a command prompt:

c:
cd C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE
cannonball-se

And paste the output. Do you have d3dcompiler_47.dll in c:\windows\system32?

If you are already compiling cannonball successfully, you could also copy oferrari.cpp from -se and that should fix your tyre slip issue.

**Reply by Beastman1976** [reply]
> Sorry this isn't working - it must be frustrating. My guess is that the graphics initialisation is failing. In VS it should have produced some console output, if not start it from a command prompt:
> 
> c: cd C:\Users\Brian\Downloads\CannonBall-SE\CannonBall-SE cannonball-se
> 
> And paste the output. Do you have d3dcompiler_47.dll in c:\windows\system32?
> 
> **If you are already compiling cannonball successfully, you could also copy oferrari.cpp from -se and that should fix your tyre slip issue.**

Could you give me a real idiots guide on how to do this please? I have visual studio installed but have no idea how to replace the oferrari.cpp file with yours in the cannonball source files, then making a new .exe file out of it.  

Thanks for your help so far. Suspect I'll need to upgrade my graphics card to get your version of the game to run properly?

**Comment by Beastman1976** [comment]
Thanks, yes I do have d3dcompiler_47.dll in c:\windows\system32

This is the output using command line:

"An enhanced build of the SEGA Outrun engine by Chris White (https://github.com/djyt/cannonball)
CannonBall-SE is Copyright (c) 2025, James Pearce (https://github.com/J1mbo/cannonball)

Automatic frame-rate selection enabled.
Unable to locate rom in path: roms/ possible name: opr-10188.71f crc32: 0xc2de09b2"

It seems to be looking for Chris White's enhanced audio rom, which I don't have. Is there any way to bypass this? I've tried copying and renaming opr-10188.71 to opr-10188.71f, but that doesn't work. I guess it's looking for that particular new sample within the rom? Can that rom be downloaded from anywhere?

**Comment by J1mbo** [comment]
Try setting "fix_samples" to 0 in config.xml

**Comment by Beastman1976** [comment]
OK almost there. I've got the game running (found the missing ROM online), and the tyre squeal issue does seem to be working correctly (yey!) However it looks like this:

![Outrun1](https://github.com/user-attachments/assets/c522701f-2b57-4e9c-a3aa-002a144a9d6f)

:(

**Reply by J1mbo** [reply]
I've seen this once on Windows also - is it the same if you restart it (the game, not the pc)?

**Comment by Beastman1976** [comment]
Ah yes it seems different every time I run the game. Took about 5 times before it came on perfect.

**Comment by Beastman1976** [comment]
![Outrun2](https://github.com/user-attachments/assets/83082d72-4074-45e6-b80c-159fb1bc5b68)

**Comment by Beastman1976** [comment]
I will have a play around with this. Thank your very much! Do you have a donations page?

**Reply by J1mbo** [reply]
Great, let me know how you get on. I'll try and hunt down that startup glitch.

And if you’re looking to donate, Chris is the one to support: https://www.patreon.com/cw/REASSEMBLER68K

**Comment by Beastman1976** [comment]
If it helps, when I go into the Blargg CRT settings and alter the saturation that changes the glitch. It can be fixed that way, but obviously you might be stuck with a saturation setting you don't want.

**Reply by J1mbo** [reply]
Yes, it really does - thanks. This should now be fixed.

Also, if the patched audio ROM isn't present, it now logs a warning and proceeds with the original also.

**Comment by Beastman1976** [comment]
Yes that is working for me now and without the patched audio rom. :) Very cool what you have done, I was having a play around with it last night. I'm a lot more motivated now to finish the game I was working on with layout. The music seems to be clearer in this version too.

I was thinking it might be nice to have a 'restore defaults' option, for people like me who like to mess up all the video settings and then can't remember what the original ones were. :)

A completely different issue I've been having is with some of the other traffic flying off screen too soon. Smashing in to me at the last moment, as if they are brake checking me. I don't know if this is a bug in the original arcade game, but I think it's probably just an issue with layout or cannonball. Seems to do it more on the stages with a raised horizon, like the gateway stage. I'll upload a video later and show you what I mean. I can live with it if it's too difficult to fix, but just thought I would mention it, as it can be annoying. The tyre squeal thing was the main one, so many thanks for fixing that :)

**Reply by J1mbo** [reply]
Restore defaults is a good shout. Right now deleting config.xml would do it, or in the menu chose "Game Modes" then "Set Enhanced Mode and Play Now" loads most settings. But agree, it could be better.

I've not seen the car behaviour described. Basically the game engine drops sprites when they're out of y clip space. This can be seen pulling away from the start line really slowly, people in the crowd just suddenly vanish. Could you post a video maybe?

**Reply by Beastman1976** [reply]
Yes, I'll try to get a video up in the next few days, been a bit busy with other things. I might send you one of my layout .bin files then you can experience it yourself. 

I've also had traffic stay in lane then fly over into my lane at the last moment and ram me off the road, but only when I'm in the adjacent lane.

**Comment by Beastman1976** [comment]
Three examples here of the cars zooming off at the end. Annoying when they hit you. It's not all cars that do it, just occasional ones:

https://github.com/user-attachments/assets/94869d3b-e5af-45da-9195-c0a1f1babb70

**Comment by Beastman1976** [comment]
Another issue I'm having with the traffic is that it is the same every game. Isn't it supposed to be randomly generated each time? At least with stage 1, I know exactly what cars will be coming up and where they will be.

**Reply by J1mbo** [reply]
That I can help you with - set randomgen to 0 in config.xml.

The traffic issue is a bit odd. It looks to me that the traffic is accelerating as it approaches the player. I'm not sure why that would happen.

**Comment by Beastman1976** [comment]
Yes that has worked thanks, I didn't know that was there in the config file. The traffic does seem to be a bit more erratic in general, with lane changing and speed changes. Although it has been a while since I played the original game to compare properly.   

Do you have any recommendations for steering wheel and pedals to use with a windows laptop for cannonball outrun?

**Comment by Beastman1976** [comment]
Something else I have noticed with my SE version of cannonball is the sound effect when overtaking traffic seems to be missing. Is there any way to get this back by altering the settings? 

Also during the animation after completing stage E, the trophy changes colour!

**Reply by J1mbo** [reply]
Interesting, I will check! I’m not sure why the sound effect is missing but will check.

I can’t think why the trophy would be different between SE and Chris’ original, is it repeatably different for that ending?

**Reply by Beastman1976** [reply]
The trophy changing colour is a bug in the original cannonball, does it every time.

**Reply by J1mbo** [reply]
Thanks. The overtake sound is fixed in the latest commit. I'd simply commented out the whole function; I have no idea why.

**Reply by Beastman1976** [reply]
Very cool, I will check it out. Thanks again, loving your work.

**Reply by Beastman1976** [reply]
Sound effects working great thanks. v1.4 seems to be a bit sluggish on my laptop though. How can I optimise the settings so that it runs a smoothly as possible? Does having too many music files make it run slower? I've only got about 7 or so, but I could get rid of a few if it will improve things.

**Reply by Beastman1976** [reply]
Theres another issue with v1.4, the traffic seems to come up out the ground? as if the wheels are buried and then as they come closer to screen they rise up to the correct level of the road. Finding myself just going back to using v1.3 and putting up with the missing sound effects. Seems to be the better version.

**Reply by J1mbo** [reply]
Is it related to hi-res sprites perhaps (settings/game-engine/enhancements, can be toggled in-game with F7)? I see it too.

**Reply by Beastman1976** [reply]
pressing F7 doesn't seem to make a difference on mine. the whole game is just more 'juddery', with the sprites not scrolling as smoothly as they should be.

**Reply by Beastman1976** [reply]
Would it be possible for you to upload a v1.3c? the same as v1.3b but with the overtake sound effects included? it was pretty much perfect. I can't tolerate v1.4, something isn't right with it.

**Reply by J1mbo** [reply]
Here you go: https://github.com/J1mbo/cannonball-se/releases/tag/v1.32

If you find it's stuttery or feels sluggish, check the vsync setting in config.xml (previously, vsync was always on regardless). Other than that, the only change is the overtake sound and things necessary to make it compile on VS'26 (which never fails to waste an hour of my day with new errors and problems!).

I will look at 1.4 at some point as there are some good improvements in that build, but I'm working on another project at the moment.

**Reply by Beastman1976** [reply]
Thankyou that is kind of you. I am happy with this 1.32 version now, it is running well for me. I just need to get my own arse into gear and finish this game I've been designing in layout. I spend more time playing it than working on it!
